OpenRA.Game/Map/CellLayer.cs

```diff
@@ -45,9 +45,23 @@ public void CopyValuesFrom(CellLayer<T> anotherLayer)
 			Array.Copy(anotherLayer.entries, entries, entries.Length);
 		}
 
+		public static CellLayer<T> CreatePrototype(Func<int, int, T> creator, int mapWidth, int mapHeight, TileShape tileShape)
+		{
+			var size = new Size(mapWidth, mapHeight);
+			var defaultCellInfoLayer = new CellLayer<T>(tileShape, size);
+			for (var v = 0; v < mapHeight; v++)
+				for (var u = 0; u < mapWidth; u++)
+					defaultCellInfoLayer[new MPos(u, v)] = creator(u, v);
+
+			return defaultCellInfoLayer;
+		}
+
 		// Resolve an array index from cell coordinates
 		int Index(CPos cell)
 		{
+			if (Shape == TileShape.Rectangle)
+				return cell.Y * Size.Width + cell.X;
```

Comment:
I realise this is for performance, but I'm not sure this justifies the code duplication.

Are there specific classes that are being slowed down significantly by this function? If so, can any of them be refactored to call the more lightweight MPos indexer and avoid the conversion cost totally?